crates/rome_js_semantic/src/events.rs

```diff
@@ -374,26 +354,45 @@ impl SemanticEventExtractor {
         debug_assert!(!self.scopes.is_empty());
 
         match self.scopes.last_mut() {
-            None => unreachable!(),
             Some(scope) => scope,
+            None => unreachable!(),
         }
     }
 
-    fn current_not_hoisting_scope_index(&mut self) -> usize {
+    /// Finds the scope where declarations that are hoisted
+    /// will be declared at. For example:
+    ///
+    /// ```js
+    /// function f() {
+    ///     if (true) {
+    ///         var a;
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// ```a``` declaration will be hoisted to the scope of
+    /// function ```f```.
+    ///
+    /// This method when called inside the ```if``` scope will return
+    /// the ```f``` scope index.
+    fn scope_index_to_hoist_declarations(&mut self) -> usize {
         // We should at least have the global scope
         // that do not hoist
-        debug_assert!(!self.scopes[0].allows_decl_hoisting);
+        debug_assert!(matches!(
+            self.scopes[0].hoisting,
+            ScopeHoisting::DontHoistDeclarationsToParent
+        ));
         debug_assert!(!self.scopes.is_empty());
 
-        let idx = self
-            .scopes
-            .iter()
-            .rev()
-            .position(|scope| !scope.allows_decl_hoisting);
+        let idx = self.scopes.iter().rev().position(|scope| {
+            matches!(scope.hoisting, ScopeHoisting::DontHoistDeclarationsToParent)
+        });
 
         match idx {
-            None => unreachable!(),
             Some(idx) => idx,
+            // Worst case this will fallback to the global scope
+            // which will be idx = 0
+            None => unreachable!(),
```

Comment:
Actually, I meant something like this:

```suggestion
            None => unreachable!("We must have a least of scope."),
```

So in the unfortunate case we reach this branch, the compiler will show us this string